core/templates/services/playthrough.service.ts

```diff
@@ -17,383 +17,268 @@
+    if (this.currentStateName === destStateName) {
+      this.numTries += 1;
+    } else {
+      this.currentStateName = destStateName;
+      this.numTries = 0;
     }
+  }
+}
 
-    private incrementIncorrectAnswerInMultipleIncorrectIssueTracker(): void {
-      PlaythroughService.multipleIncorrectStateName.num_times_incorrect += 1;
-    }
+class CyclicStateTransitionsTracker {
+  public visitedStates: string[];
+  public cycle: string[] = null;
+  public numCycles: number = 0;
 
-    private recordStateTransitionInMultipleIncorrectIssueTracker(
-        destStateName: string): void {
-      if (PlaythroughService.multipleIncorrectStateName.num_times_incorrect <
-        ServicesConstants.NUM_INCORRECT_ANSWERS_THRESHOLD) {
-        PlaythroughService.multipleIncorrectStateName.state_name = (
-          destStateName);
-        PlaythroughService.multipleIncorrectStateName.num_times_incorrect = 0;
-      }
-    }
+  constructor(initStateName: string) {
+    this.visitedStates = [initStateName];
+  }
 
-    private recordStateTransitionInCyclicIssueTracker(
-        destStateName: string): void {
-      if (PlaythroughService.cycleIdentifier.num_cycles <
-        ServicesConstants.NUM_REPEATED_CYCLES_THRESHOLD) {
-        if (PlaythroughService.visitedStates.indexOf(destStateName) !== -1) {
-          // Cycle identified.
-          let cycleStartIndex = PlaythroughService.visitedStates.indexOf(
-            destStateName);
-          PlaythroughService.visitedStates.push(destStateName);
-          let cycleString = PlaythroughService.visitedStates.slice(
-            cycleStartIndex, PlaythroughService.visitedStates.length
-          ).toString();
-          if (PlaythroughService.cycleIdentifier.cycle === cycleString) {
-            PlaythroughService.cycleIdentifier.num_cycles += 1;
-          } else {
-            PlaythroughService.cycleIdentifier.cycle = cycleString;
-            PlaythroughService.cycleIdentifier.num_cycles = 1;
-          }
-          PlaythroughService.visitedStates = [destStateName];
-        } else {
-          PlaythroughService.visitedStates.push(destStateName);
-        }
-      }
-    }
+  private makeCycle(cycleStartIndex: number): string[] {
+    const collision = this.visitedStates[cycleStartIndex];
+    const cycleWithoutCollision = this.visitedStates.slice(cycleStartIndex);
+    return [...cycleWithoutCollision, collision];
+  }
 
-    private isMultipleIncorrectSubmissionsIssue(): boolean {
-      return (
-        PlaythroughService.multipleIncorrectStateName.num_times_incorrect >=
-        ServicesConstants.NUM_INCORRECT_ANSWERS_THRESHOLD);
-    }
+  private currentStateName(): string {
+    return this.visitedStates[this.visitedStates.length - 1];
+  }
 
-    private isCyclicStateTransitionsIssue(): boolean {
-      return PlaythroughService.cycleIdentifier.num_cycles >=
-        ServicesConstants.NUM_REPEATED_CYCLES_THRESHOLD;
-    }
+  isIssue(): boolean {
```

Comment:
I thought "issue" already meant something else in the codebase (i.e. collection of playthroughs)? Maybe use a different name to avoid ambiguity... (isExampleOfIssue or isProblematic() or isEligibleForRecording perhaps)?